### ISSUE TITLE
Support cycling logs Was: Ban after too few attempts due to erroneous log rotation detection

### ISSUE THREAD

**Issue by cepheid666** [fail2ban/fail2ban#17]
(Posted here based on discussion through the fail2ban-users mailing list)

I have fail2ban monitoring a couple of BB error logs that are causing some havok with its logfile tracking.  Specifically, the YaBB error logs aren't arbitrary-length, constant-append logs like most error logs (e.g. apache error log), but are instead fixed-length, constantly-overwritten logs - they are capped at N lines (say N = 100) and every new (last) line written to the file causes the oldest (first) line to be deleted.

Fail2ban's log rotation detection is barfing on this kind of log - it thinks the log is being rotated every time it sees a new line being written.  As a result, it basically "rescans" the file every time, but without resetting the hack-attempt counter, causing each line to be counted multiple times.

As an example:
Say the maxretry for this jail is set to 10 attempts.  Fail2ban notices the first matching hack attempt in the error log, and sets the counter to 1.  A new hack attempt is written to the file; upon the next scan, fail2ban thinks the logfile got rotated, and rescans the entire file - it now sees 2 entries, and adds BOTH entries to the counter, which is now at 3 attempts.  The same thing happens on the 3rd attempt, whereupon the counter is now set to 6, and for the 4th, where the counter now hits 10 and fail2ban locks out the offending IP.  In this way, fail2ban thinks it sees 10 matching lines even though it actually saw only 4, because it rescanned the logfile after each "rotation" without resetting the counter.

The one obvious workaround for this type of file is to set maxretry to the correct "n-sum" number, where if n is the number of attempts we actually want to catch, we set maxretry = n + (n-1) + (n-2) + ... + 1.  This is rather a poor hack but it'd work.

One possible solution is to disable logrotation detection for this type of log, whereupon fail2ban would ignore logrotation and simply look at the latest lines being written.  Unfortunately, since the file is of "constant" length, it would be difficult to keep track of how many new lines were added and which ones they are.

Another potential (possibly best) solution is to reset the attempt counter when log rotation is "detected" for this type of log; thus, when fail2ban rescans the entire logfile after "rotation," it will still see the correct number of matching errors instead of accumulating them the way it does now.  The reset action would need to be a user-settable flag for only this type of logfile, since it would not be desirable to do such a reset on regular rotating logs; thus, one could add a flag such as "cyclinglog = true" to the jail configuration, and if this flag is true, the attempt counter is reset upon rotation detection, but if false (default), everything would continue as normal.

It's a rare use case but IMHO should be treated.

Thanks!

**Comment by mcassaniti** [comment]
I think I am hitting this issue. @sebres you might be able to tell better than I can. I am truncating a log on a timer and removing old values. The log is __not__ empty after it is truncated. I see "Detected a log entry 54m 49s before the current time in operation mode. This looks like a latency problem. Treating such entries as if they just happened." I'm using inotify for the polling mode. All suggestions are welcome.

**Comment by sebres** [comment]
Well it can be true - the detection of log rotation is based on few things:
- file disappears for few moments (or e. g. received a deletion/renaming event from inotify)
- and/or inode of log-file gets changed
- and/or size of the log-file is smaller than previously known
- and/or md5 of first row in log is different

I cannot estimate what happens without to investigate deeper and probably without more details, but...
The message "Detected a log entry ..." occurs if fail2ban find a message with a timestamp much smaller than current time, what is imaginable by such "rotation" by truncating from head - all of the messages in log will be recognized as they happen right now (in operation mode), so fail2ban may begin to process them again.

This is indeed not quite correct way to do the rotation. Proper way would be to rename or delete previous log-file and create new empty log with the same name and permissions.

There is also no option at the moment to signal jail about such strange rotation, keeping part of old messages.
However I don't know the way how one can implement that (fail2ban needs to seek to end of log after such rotation), because trimming from ahead is not atomic operation and fail2ban would then need to wait until it is completed to ignore all that messages, before seeks to the end.

**Comment by mcassaniti** [comment]
So what happens if the log is deleted and a new log is created and populated with the same name?

**Comment by sebres** [comment]
> So what happens if the log is deleted and a new log is created and populated with the same name?

Obviously, since this is the way how all known log-rotation mechanics do, jail/backend would work as expected - it'd detect a rotation and start to monitor new log-file from scratch.

Additionally important is:
- it'd not get a bunch of old messages from old (rotated) log-file. Only new messages should appear there.
- the application also knows about new log and writes there, so not continues to log into old file (if the rotation is done outside of application, mostly it'd be notified via `postrotate` command)

Read about logrotation of Linux distro for more details.

**Comment by mcassaniti** [comment]
I'll re-write. Wasn't sure if it needed some sort of additional notification. Too easy.
